benchmarks/instruction_counts/execution/work.py

```diff
@@ -0,0 +1,204 @@
+"""Handle the details of subprocess calls and retries for a given benchmark run."""
+import dataclasses
+import os
+import pickle
+import signal
+import subprocess
+import time
+from typing import List, Optional, Union, TYPE_CHECKING
+import uuid
+
+from core.api import AutoLabels
+from core.types import Label
+from core.utils import get_temp_dir
+from worker.main import WORKER_PATH, WorkerFailure, WorkerOutput, WorkerTimerArgs, WorkerUnpickler
+
+if TYPE_CHECKING:
+    PopenType = subprocess.Popen[bytes]
+else:
+    PopenType = subprocess.Popen
+
+
+# Mitigate https://github.com/pytorch/pytorch/issues/37377
+_ENV = "MKL_THREADING_LAYER=GNU"
+_PYTHON = "python"
+PYTHON_CMD = f"{_ENV} {_PYTHON}"
+
+# We must specify `bash` so that `source activate ...` always works
+SHELL = "/bin/bash"
+
+
+@dataclasses.dataclass(frozen=True)
+class WorkOrder:
+    """Spec to schedule work with the benchmark runner."""
+    label: Label
+    autolabels: AutoLabels
+    timer_args: WorkerTimerArgs
+    source_cmd: Optional[str] = None
+    timeout: Optional[float] = None
+    retries: int = 0
+
+    def __hash__(self) -> int:
+        return id(self)
+
+
+class _BenchmarkProcess:
+    """Wraps subprocess.Popen for a given WorkOrder."""
+    _work_order: WorkOrder
+    _cpu_list: Optional[str]
+    _proc: PopenType
+
+    # Internal bookkeeping
+    _communication_file: str
+    _start_time: float
+    _end_time: Optional[float] = None
+    _retcode: Optional[int]
+    _result: Optional[Union[WorkerOutput, WorkerFailure]] = None
+
+    def __init__(self, work_order: WorkOrder, cpu_list: Optional[str]) -> None:
+        self._work_order = work_order
+        self._cpu_list = cpu_list
+        self._start_time = time.time()
+        self._communication_file = os.path.join(get_temp_dir(), f"{uuid.uuid4()}.pkl")
+        with open(self._communication_file, "wb") as f:
+            pickle.dump(self._work_order.timer_args, f)
+
+        self._proc = subprocess.Popen(
+            self.cmd,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            shell=True,
+            executable=SHELL,
+        )
+
+    def clone(self) -> "_BenchmarkProcess":
+        return _BenchmarkProcess(self._work_order, self._cpu_list)
+
+    @property
+    def cmd(self) -> str:
+        cmd: List[str] = []
```

Comment:
Similar question as above, maybe we should return both command and environment here, so that we can set the "GOMP_CPU_AFFINITY" environment variable as well?

Reply:
Indeed.